Fix random polynomial bias

grhkm21 opened the pull request:
For a polynomial ring `R` say `GF(11)["x"]`, before `R.random_element` is very far from random, now it is uniformly random.

grhkm21 committed:
Fix random polynomial bias

grhkm21 committed:
Allow `degree=-1` together with `monic=True`

It is more useful in practice

grhkm21 committed:
Fix buggy sampling

tscrim commented on src/sage/rings/polynomial/polynomial_ring.py:
```diff
@@ -1370,6 +1370,10 @@ def random_element(self, degree=(-1,2), *args, **kwds):
 
+        -  ``monic`` - optional boolean to indicate whether the sample
+           polynomial should be monic or not. If degree includes -1, it is still
+           possible for the algorithm to return ``0``, which is not monic.
```
This is not quite correct. Indeed, we do not consider `0` to be monic, but it is $1 \cdot 0$, where $0$ is the "basis" element of degree -1. I think you should simply adjust the degree to be nonnegative. It is valid input and the output matches it, provided `is_monic()` is always `True`.

grhkm21 replied:
I don’t understand this comment. So what is your suggestion? That if monic is set then degree can’t be -1? (I think it’s commented out below, can’t really check on phone.)

I decided against since for my own use case, I actually want to sample all polynomials with degree -1 to k, and if it is not degree -1, then make it monic. I think in general this use case is common too, and if this is undesirable ie degree -1 shouldn’t be returned, then the user can just write degree=(0,k)

tscrim replied:
Allowing degree -1 is fine, but when `monic=True`, it should never return `0` as that is not a monic polynomial (by Sage's definition/convention). Basically, you should always have `P.random_polynomial(monic=True).is_monic()` return `True` no matter what the degree range is.

Thus within the code, it should simply (silently) have `degree[0]` become `max(0, degree[0])` when `monic=True`.

Note that, e.g., `degree=(-5, 10)` should be considered as valid input IMO (although it currently throws an error). Nothing requires that the result be uniform, much less have full support of the input.

grhkm21 replied:
> Allowing degree -1 is fine, but when `monic=True`, it should never return `0` as that is not a monic polynomial (by Sage's definition/convention). Basically, you should always have `P.random_polynomial(monic=True).is_monic()` return `True` no matter what the degree range is.

I have reworded `monic` to `monic_or_zero`. If the user doesn't want zero, they can set `degree[0] = 0`, or call `random_monic_element(zero=False)`. I prefer giving user freedom when they *do* want zero.

> Note that, e.g., `degree=(-5, 10)` should be considered as valid input IMO (although it currently throws an error). Nothing requires that the result be uniform, much less have full support of the input.

I wonder if that'll be more vulnerable to errors, but I'll apply that.

tscrim commented on src/sage/rings/polynomial/polynomial_ring.py:
```diff
@@ -1448,24 +1466,64 @@ def random_element(self, degree=(-1,2), *args, **kwds):
-        # Pick a random degree
-        d = randint(degree[0], degree[1])
+        if degree == (-1, -1):
+            return 0
```
```suggestion
        if degree == (-1, -1):
            return self.zero()
```
The Python `int` `0` is not the same as the zero in the polynomial ring; it has different methods.

tscrim commented on src/sage/rings/polynomial/polynomial_ring.py:
```diff
@@ -1448,24 +1466,64 @@ def random_element(self, degree=(-1,2), *args, **kwds):
+
+        while True:
+            # Pick random coefficients
+            coefs = []
```
Rather than constantly appending to a list, it is faster to populate the list with `None` and then modify the entries. Since you know exactly how many elements will be in the list, you can do this.

grhkm21 committed:
Merge branch 'develop' into grhkm/uniform-polynomial-ring

grhkm21 committed:
apply review changes 📖

grhkm21 committed:
apply review changes 📖

tscrim commented on src/sage/rings/polynomial/polynomial_ring.py:
```diff
@@ -1360,23 +1360,41 @@ def ngens(self):
+
         -  ``*args, **kwds`` - Passed on to the ``random_element`` method for
-           the base ring
+           the base ring.
```
```suggestion
           the base ring
```
Revert; this is against our doc style.

tscrim commented on src/sage/rings/polynomial/polynomial_ring.py:
```diff
@@ -1360,23 +1360,41 @@ def ngens(self):
 
         -  ``degree`` - optional integer for fixing the degree
            or a tuple of minimum and maximum degrees. By default set to
            ``(-1,2)``.
```
```suggestion
        -  ``degree`` -- (default: ``(-1, 2)``) integer for fixing the degree
           or a tuple of minimum and maximum degrees
```
Let's make the doc into the correct format while modifying it.

grhkm21 replied:
Is the correct format to not include a full stop at the end of "INPUT:" docstrings?

tscrim replied:
Generally yes. The exception is if you need multiple sentences to describe the input (although there are other ways to do more complicated behaviors).

tscrim commented on src/sage/rings/polynomial/polynomial_ring.py:
```diff
@@ -1360,23 +1360,41 @@ def ngens(self):
+        -  ``monic`` - optional boolean to indicate whether the sampled
+           polynomial should be monic, or not. If this is set, `0` is not a
+           possible output of the method. If monic and `0` are both needed,
+           one should use :meth:`random_monic_or_zero_element`.
```
```suggestion
        -  ``monic`` -- boolean  (default: ``False``); whether the sampled
           polynomial should be monic, or not
```
See topline review comment.

tscrim commented on src/sage/rings/polynomial/polynomial_ring.py:
```diff
@@ -1360,23 +1360,41 @@ def ngens(self):
             sage: R.<x> = ZZ[]
-            sage: f = R.random_element(10, 5, 10)
+            sage: f = R.random_element(10, x=5, y=10)
+            sage: f # random
```
```suggestion
            sage: f  # random
```
PEP8 thing IIRC; at least we are doing this spacing uniformly. Although rather than having the random polynomial, it is better to check the properties.

tscrim reviewed (changes requested):
This is now even more complicated to make something that is actually wrong be a returned value. Again, the zero polynomial is **not** monic. It should **not** be returned when `monic=True`, ever. It is a contradiction. Remove the `_allow_zero`, then apply fire to it. Then repeat for everything related to this.

Additionally, I see no point in having other methods that simply are setting options in another method. That is why we have arguments, right? These methods are creating extra maintenance complications and polluting tab completion IMO.

grhkm21 committed:
apply review changes 📖

This commit removes all `monic_or_zero` methods.

grhkm21 committed:
fix docstring fullstops

grhkm21 committed:
fix styling

tscrim commented on src/sage/rings/polynomial/polynomial_ring.py:
```diff
@@ -1360,23 +1360,32 @@ def ngens(self):
+           possible output of the method
 
         -  ``*args, **kwds`` - Passed on to the ``random_element`` method for
            the base ring
```
```suggestion
        -  ``degree`` — (default: ``(-1, 2)``) integer for fixing the degree or
           a tuple of minimum and maximum degrees

        -  ``monic`` — boolean (default: ``False``) whether the sampled
           polynomial should be monic

        -  additional (keyword) arguments are passed on to the ``random_element``
           method for the base ring
```
Since `0` is not a monic polynomial, it doesn’t need to be stated that it won’t return `0`.

tscrim commented on src/sage/rings/polynomial/polynomial_ring.py:
```diff
@@ -1448,24 +1481,101 @@ def random_element(self, degree=(-1,2), *args, **kwds):
+        # If `monic` is set, zero should be ignored
+        if degree[0] == -1:
+            degree = (0, degree[1])
+            has_zero = not monic
```
The `has_zero` variable is not serving any purpose anymore and can be removed in favor of `monic`. This will also simplify the non-monic logic.

tscrim commented on src/sage/rings/polynomial/polynomial_ring.py:
```diff
@@ -1448,24 +1481,101 @@ def random_element(self, degree=(-1,2), *args, **kwds):
+                continue
 
-        return p
+            coefs.reverse()
```
Rather than reversing here, shouldn’t you just construct the list in the correct order?

tscrim commented on src/sage/rings/polynomial/polynomial_ring.py:
```diff
@@ -1421,10 +1445,17 @@ def random_element(self, degree=(-1,2), *args, **kwds):
-            Traceback (most recent call last):
-            ...
-            ValueError: degree should be an integer greater or equal than -1
+        In :issue:`37118`, ranges including integers below `-1` no longer error::
```
```suggestion
        In :issue:`37118`, ranges including integers below `-1` no longer
        raise an error::
```

tscrim commented on src/sage/rings/polynomial/polynomial_ring.py:
```diff
@@ -1401,6 +1410,21 @@ def random_element(self, degree=(-1,2), *args, **kwds):
+            sage: R.random_monic_element(degree=(-1, 5)).is_monic()
+            True
+
+        Note that if the degree range includes `-1` and ``monic`` is set, it is silently ignored::
```
This could be changed to remind the user that `0` is not a monic polynomial.

tscrim reviewed (changes requested):
Thanks. I am glad things are clarified.

Let me remind you about my comment that having methods that simply call another with a default is not good practice. You already have the argument, it makes the code more complicated with more to maintain (and potential to cause conflicts), and pollutes tab completion. `random_monic_polynomial` should be removed.

You also need to include a check if `monic=True` and the only degree(s) possible are `-1`. Related, if `degree[1] < -1` then you also need to error out.

I don’t find it useful to have a `# random` output, but instead to test the desired properties of the output polynomial (e.g., the degree is within the correct range).

grhkm21 committed:
apply review changes 📖

grhkm21 committed:
apply review changes 📖

tscrim commented on src/sage/rings/polynomial/polynomial_ring.py:
```diff
@@ -1401,6 +1405,14 @@ def random_element(self, degree=(-1,2), *args, **kwds):
             ....:     pass
 
+        Note that if the degree range includes `-1` and ``monic`` is set, it is
+        silently ignored, as `0` is not a monic polynomial::
```
```suggestion
        Monic polynomials are also chosen among all monic polynomials
        with degree between the given  ``degree`` argument::
```
I don't think you need to tell people that `0` is not a monic polynomial. :) Same comment about uniformity.

tscrim commented on src/sage/rings/polynomial/polynomial_ring.py:
```diff
@@ -1433,11 +1463,17 @@ def random_element(self, degree=(-1,2), *args, **kwds):
         if degree[0] <= -2:
-            raise ValueError("degree should be an integer greater or equal than -1")
+            # This error has been removed in issue #37118.
+            # raise ValueError("degree should be an integer greater or equal than -1")
```
```suggestion
```
The comment is not necessary. We have the git history.

tscrim commented on src/sage/rings/polynomial/polynomial_ring.py:
```diff
@@ -1448,24 +1484,48 @@ def random_element(self, degree=(-1,2), *args, **kwds):
-        return p
+        # If `monic` is set, zero should be ignored
+        if degree[0] == -1 and monic:
+            degree = (0, degree[1])
```
```suggestion
        if degree[0] == -1 and monic:
            if degree[1] == -1:
                raise ValueError("the maximum degree of monic polynomials needs to be at least 0")
            if degree[1] == 0:  # Must be a constant and monic, so 1
                return self.one()
            degree = (0, degree[1])
```

tscrim commented on src/sage/rings/polynomial/polynomial_ring.py:
```diff
@@ -1360,23 +1360,27 @@ def ngens(self):
-        -  ``degree`` - optional integer for fixing the degree
-           or a tuple of minimum and maximum degrees. By default set to
-           ``(-1,2)``.
+        -  ``degree`` - (default: ``(-1, 2)``) integer for fixing the degree or
```
```suggestion
        -  ``degree`` -- (default: ``(-1, 2)``) integer for fixing the degree or
```

tscrim commented on src/sage/rings/polynomial/polynomial_ring.py:
```diff
@@ -1360,23 +1360,27 @@ def ngens(self):
+           polynomial should be monic
+
+        -  ``*args, **kwds`` - additional keyword parameters passed on to the
+           ``random_element`` method for the base ring
```
```suggestion
        -  ``monic`` -- boolean (optional); indicate whether the sampled
           polynomial should be monic

        -  ``*args, **kwds`` -- additional keyword parameters passed on to the
           ``random_element`` method for the base ring
```

tscrim commented on src/sage/rings/polynomial/polynomial_ring.py:
```diff
@@ -1448,24 +1484,48 @@ def random_element(self, degree=(-1,2), *args, **kwds):
-
-        # If degree is -1, return the 0 polynomial
-        if d == -1:
+        degree_lb = degree[0]
```
```suggestion
```

tscrim commented on src/sage/rings/polynomial/polynomial_ring.py:
```diff
@@ -1448,24 +1484,48 @@ def random_element(self, degree=(-1,2), *args, **kwds):
+            # Pick random coefficients
+            end = degree[1]
+            coefs = [None] * (degree[1] + 1)
+            nonzero = False
```
```suggestion
        end = degree[1]
        if degree[0] == -1:  # already know it is not monic
            # Since we allow 0 as an output, we can get just a random list of elements
            return self([R.random_element(*args, **kwds) for _ in range(end+1)])

        nonzero = False
        coefs = [None] * (end + 1)  # We will always replace every element each iteration
        while not nonzero:
            # Pick random coefficients
```

tscrim commented on src/sage/rings/polynomial/polynomial_ring.py:
```diff
@@ -1448,24 +1484,48 @@ def random_element(self, degree=(-1,2), *args, **kwds):
+            if degree_lb == -1 and not nonzero:
+                continue
+
+            return self(coefs)
```
```suggestion
        # Now we pick the remaining coefficients
        for i in range(degree[1] - degree[0] + 1, degree[1] + 1):
            coefs[end - i] = R.random_element(*args, **kwds)

        return self(coefs)
```

tscrim commented on src/sage/rings/polynomial/polynomial_ring.py:
```diff
@@ -1386,8 +1390,8 @@ def random_element(self, degree=(-1,2), *args, **kwds):
-        If a tuple of two integers is given for the ``degree`` argument, a degree
-        is first uniformly chosen, then a polynomial of that degree is given::
+        If a tuple of two integers is given for the ``degree`` argument, a polynomial is chosen
+        uniformly among all polynomials with degree between them::
```
```suggestion
        among all polynomials with degree between them::
```
I am dubious about this uniform distribution claim since it depends highly on the sampling of the coefficient ring (which doesn't have to be uniform).

grhkm21 replied:
I think the information is still important to note for cases like $\mathbb{F}_{103}[x]$ which is uniform. I now phrase it as

```
If a tuple of two integers is given for the ``degree`` argument, a polynomial is chosen
among all polynomials with degree between them. If the base ring is uniform, so is this
method::
```

is that ok? Not sure if the English is really correct.

tscrim replied:
I think the last sentence needs to be something like "If the base ring is sampled uniformly, then this is also sampled uniformly::"

I am wondering if we should include a test about the uniformness. It's hard to test it reliably is random, but perhaps we should sample over something like 10000 polynomials of degree 3 over $F_3$ and check the distribution for the first digit? I don't think this is enough for 2 digits in general. It might not be worth it to spend a lot of time on this.

grhkm21 replied:
I agree that we should have a test for uniformness, but I don't know how to approach this correctly. I tried a chi-squared test with say $p = 0.05$, but it often fails, c.f. ["Too Big to Fail: Large Samples and the p-Value Problem"](https://www.researchgate.net/publication/270504262_Too_Big_to_Fail_Large_Samples_and_the_p-Value_Problem).

tscrim replied:
Yea, I couldn't think of a good way to reasonably guarantee this. Let's just leave it.

However, it doesn't make sense to say "the base ring is uniform". You need to discuss about the sampling. My previous suggestion can also be improved by "If we can sample uniformly from the base ring, then ..."

tscrim replied:
This still needs to be addressed. It does not make sense to say the base ring is uniform.

grhkm21 replied:
Oh god, I completely missed this comment. I have addressed it now though the wording is slightly different. (I fixed it before I saw your newest comment two days ago)

tscrim reviewed (changes requested):
No problem. Thank you for the additional changes.

However, I do feel somewhat strongly about not having the `# random` tests here. Having the output there is useless IMO and could even potentially confuse a user. They were not included here for a reason I believe.

I also have some additional comments to try and improve the code some more.

grhkm21 committed:
Merge branch 'develop' into grhkm/uniform-polynomial-ring

grhkm21 committed:
refactor `.random_element` 📖

grhkm21 committed:
fix input docs format

grhkm21 committed:
I don't know what the code is trying to do here

grhkm21 committed:
fix remaining doctests

grhkm21 committed:
fix docs wording + uniformity test 📖

grhkm21 committed:
Fix doctests

As I noted in the discussion under :issue:`37118`, it seems that the
output of `sage -t` and the REPL differ, so I am not sure if this commit
is correct.

grhkm21 committed:
long time doctests

grhkm21 committed:
Merge branch 'develop' into grhkm/uniform-polynomial-ring

grhkm21 committed:
Small wording change

tscrim commented on src/sage/rings/polynomial/polynomial_ring.py:
```diff
@@ -2414,8 +2477,8 @@ def lagrange_polynomial(self, points, algorithm="divided_difference", previous_r
-        # using Neville's method for recursively generating the
-        # Lagrange interpolation polynomial
+# using Neville's method for recursively generating the
+# Lagrange interpolation polynomial
```
Please revert. The indentation is useful and such a change can just create (merge) conflicts.

grhkm21 replied:
Sorry, I blame this on my editor.

tscrim reviewed (approved):
Thank you. One last little thing to do. It would also be good if the lines you changed were at 80 char/line or less too.

grhkm21 committed:
wrap some lines fix some indentation

grhkm21 committed:
wrap at 80
